sotodlib/io/bookbinder2.py

```diff
@@ -0,0 +1,931 @@
+        if 'Corrected_Corotation' in self.data:
+            corotation = np.interp(times, self.times, self.data['Corrected_Corotation'])
+
+        anc_frame_data = []
+        for oframe_idx in np.unique(frame_idxs):
+            if file_idxs[oframe_idx] != cur_file_idx:
+                close_writer(writer)
+                cur_file_idx = file_idxs[oframe_idx]
+                fname = os.path.join(outdir, f'A_ancil_{cur_file_idx:0>3}.g3')
+                out_files.append(fname)
+                writer = core.G3Writer(fname)
+
+            m = frame_idxs == oframe_idx
+            ts = times[m]
+
+            oframe = core.G3Frame(core.G3FrameType.Scan)
+
+            i0, i1 = np.where(m)[0][[0, -1]]
+            oframe['sample_range'] = core.G3VectorInt([int(i0), int(i1+1)])
+            oframe['book_id'] = self.book_id
+
+            anc_data = core.G3TimesampleMap()
+            anc_data.times = core.G3VectorTime(ts * core.G3Units.s)
+            if az is not None:
+                anc_data['az_enc'] = core.G3VectorDouble(az[m])
+                anc_data['el_enc'] = core.G3VectorDouble(el[m])
+            if boresight is not None:
+                anc_data['boresight_enc'] = core.G3VectorDouble(boresight[m])
+            if corotation is not None:
+                anc_data['corotation_enc'] = core.G3VectorDouble(corotation[m])
+            oframe['ancil'] = anc_data
+            writer(oframe)
+            anc_frame_data.append(anc_data)
+
+        # Save this to be added to detector files
+        self.anc_frame_data = anc_frame_data
+        self.out_files = out_files
+        
+
+class SmurfStreamProcessor:
+    def __init__(self, obs_id, files, book_id, readout_ids=None, log=None):
+        self.files = files
+        self.obs_id = obs_id
+        self.stream_id = None
+        self.times = None
+        self.frame_idxs = None
+        self.nchans = None
+        self.nframes = None
+        self.bias_names = None
+        self.primary_names = None
+        self.timing_paradigm = None
+        self.session_id = None
+        self.slow_primary = None
+        self.sostream_version = None
+        self.readout_ids = readout_ids
+        self.out_files = []
+        self.book_id = book_id
+
+        if log is None:
+            self.log = logging.getLogger('bookbinder')
+        else:
+            self.log = log
+
+    def preprocess(self):
+        """
+        Extracts file times, nchans, and nnframes from file list
+        """
+        if self.times is not None:  # Already preprocessed
+            return
+
+        self.log.info(f"Preprocessing smurf obsid {self.obs_id}")
+
+        self.nframes = 0
+        ts = []
+        frame_idxs = []
+        frame_idx = 0
+        for frame in get_frame_iter(self.files):
+            if frame.type != core.G3FrameType.Scan:
+                continue
+
+            if self.readout_ids is None:
+                self.readout_ids = frame['data'].names
```

Comment:
unless I'm reading this wrong, we don't want this fall back. If we can't find readout id's through g3tsmurf we want to throw errors